Source/DafnyLanguageServer/Workspace/DocumentManager.cs

```diff
@@ -0,0 +1,180 @@
+    var updatedText = textChangeProcessor.ApplyChange(CompilationManager.TextBuffer, documentChange, CancellationToken.None);
+
+    var lastPublishedDocument = observer.LastPublishedDocument;
+    var oldVerificationDiagnostics = lastPublishedDocument.ImplementationIdToView;
+    var migratedImplementationViews =
+      oldVerificationDiagnostics == null ? null : MigrateImplementationViews(documentChange, oldVerificationDiagnostics);
+
+    // TODO it would be simpler to store last touched positions at the DocumentState level, and always recompute the lastTouchedVerifiables from that.
+    // Then we don't need to touch lastPublishedDocument.LastTouchedVerifiables which seems unreliable.
+    var migratedLastTouchedVerifiables =
+      relocator.RelocatePositions(lastPublishedDocument.LastTouchedVerifiables, documentChange, CancellationToken.None);
+
+    var migratedVerificationTree =
+      relocator.RelocateVerificationTree(lastPublishedDocument.VerificationTree, updatedText.NumberOfLines, documentChange, CancellationToken.None);
+
+    var lastChange = documentChange.ContentChanges.Select(contentChange => contentChange.Range).LastOrDefault();
+
+    CompilationManager = new CompilationManager(
+      services,
+      verifierOptions,
+      updatedText,
+      lastChange,
+      VerifyOnChange,
+      migratedLastTouchedVerifiables,
+      // TODO do not pass this to CompilationManager but instead use it in FillMissingStateUsingLastPublishedDocument
+      migratedVerificationTree
+    );
+
+    observerSubscription.Dispose();
+    var migratedUpdates = CompilationManager.DocumentUpdates.Select(document =>
+      FillMissingStateUsingLastPublishedDocument(documentChange, document, lastPublishedDocument, migratedImplementationViews));
+    observerSubscription = migratedUpdates.Subscribe(observer);
+  }
+
+  private DafnyDocument FillMissingStateUsingLastPublishedDocument(DidChangeTextDocumentParams documentChange,
+    DafnyDocument document, DafnyDocument lastPublishedDocument,
+    IReadOnlyDictionary<ImplementationId, ImplementationView>? migratedImplementationViews) {
+    if (!document.SymbolTable.Resolved) {
+      document = document with {
+        SymbolTable =
+        relocator.RelocateSymbols(lastPublishedDocument.SymbolTable, documentChange, CancellationToken.None)
+      };
+    }
+
+    var migratedViews = document.ImplementationIdToView?.Select(kv => {
+      var value = kv.Value.Status < PublishedVerificationStatus.Error
+        ? kv.Value with {
+          Diagnostics = migratedImplementationViews != null &&
+                        migratedImplementationViews.TryGetValue(kv.Key, out var previousView)
+            ? previousView.Diagnostics
+            : kv.Value.Diagnostics
+        }
+        : kv.Value;
+      return new KeyValuePair<ImplementationId, ImplementationView>(kv.Key, value);
+    }) ?? migratedImplementationViews;
+    document.ImplementationIdToView = migratedViews == null ? null : new(migratedViews);
```

Comment:
It would be nice to simplify that code by using empty dictionaries instead of `null`, if possible.

Reply:
Right now a document that has been resolved will use the migrated `ImplementationIdToView` from the previous document version without modifications, while a document that has been translated and has just created its own `ImplementationIdToView`, will combine this with the migrated one in a specific way. Using this combine function with an empty dictionary and a migrated one would give bad results.

In the future I would like to have types for the specific phases a DafnyDocument can be in, so one `TranslatedDocument` type, for which there's be a non-null field `ImplementationIdToView`, and a `ResolvedDocument` type which won't have the `ImplementationIdToView` field.